File: src/main/scala/com/okosmos/camino/domain/Position.scala
package com.okosmos.camino.domain

trait Position {
  def x: Int
  def y: Int
  def nextMoves: Seq[Position]
  override def toString: String = s"Position(${x+1},${y+1})"
}

case class KnightPosition(x: Int, y: Int) extends Position {
  def nextMoves: Seq[KnightPosition] = {
    val signs = Seq(-1, 1)
    val moves = Seq(1, 2)
    for {
      xSign <- signs
      ySign <- signs
      xDelta <- moves
      yDelta <- moves diff Seq(xDelta)
    } yield KnightPosition(x + (xSign * xDelta), y + (ySign * yDelta))
  }
  override def toString: String = s"KnightsPosition(${x+1},${y+1})"
}

File: src/main/java/com/okosmos/camino/domain/Position.java
package com.okosmos.camino.domain;

import java.util.List;

public abstract class Position {

    abstract int getX();

    abstract int getY();

    abstract List<Position> nextMoves();

    @Override
    public String toString() {
        return "Position(" + getX() + "," + getY() + ")";
    }

}
